[FIX] mrp: do not update consumed qty if manually set

lase-odoo opened the pull request:
### Steps to reproduce:

- Create a product FP with a BOM consuming 1 unit of a product COMP
- Create and confirm a MO for 1 unit of FP using the BOM
- Set the consumed quantity of COMP to 2 and tick consumed
- Save and "Produce All"

#### > the consumed quantity of COMP was updated to 1 according to the BOM

### Cause of the issue:

Clicking on "Produce All" will trigger the `pre_button_mark_done` that will update the qty_producing according to the bom proportions because of these lines:
https://github.com/odoo/odoo/blob/86f977f8f4b9859790ad77af3a40234c8eb374e9/addons/mrp/models/mrp_production.py#L1217-L1227 However, this update should be bypassed if the quantity was set on the move. This is the role of these lines:
https://github.com/odoo/odoo/blob/86f977f8f4b9859790ad77af3a40234c8eb374e9/addons/mrp/models/mrp_production.py#L1218-L1220

The issue does not appear in 17.0 because, thanks to commit https://github.com/odoo/odoo/commit/5bb0f96f1973fa7e19b6701944b4a29577f3314f the `manual_consumption` field of the stock moves related to a manufacturing order is set to be True as soon as the quantity is changed because of these lines:
https://github.com/odoo/odoo/blob/9a11717c17b860ec2f1b2e228517c0d3945c474a/addons/mrp/static/src/widgets/mrp_consumed.js#L25-L27 However, this js file was removed in saas-17.2 by commit https://github.com/odoo/odoo/commit/e143345193577442ffd1f46272a09d7a5801d567.

opw-3934942
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/166458).

adwid commented on addons/mrp/tests/test_manual_consumption.py:
```diff
@@ -230,3 +230,21 @@ def get_moves(mo):
+        Check that the manual consumption is set to true when the quantity is manualy set.
+        """
+        mo_form = Form(self.env['mrp.production'])
+        mo_form.product_id = self.bom_3.product_id
```
It's a bit weird to use the product of bom 3 while you're using bom 1
Also, setting bom 1 on the form will define the finished product
```suggestion
```

lase-odoo replied:
My bad ! Good catch :)

adwid commented on addons/mrp/tests/test_manual_consumption.py:
```diff
@@ -230,3 +230,21 @@ def get_moves(mo):
+            with fmo.move_raw_ids.edit(0) as line_0:
+                line_0.quantity = 3.0
+                line_0.picked = True
+        mo = fmo.save()
```
```suggestion
```
This will be already one when exiting the `with`
https://github.com/odoo/odoo/blob/4251995c04056a71c60cd39dbd5aff22d308ecd1/odoo/tests/form.py#L400-L402

adwid commented on addons/mrp/models/stock_move.py:
```diff
@@ -274,6 +274,12 @@ def _onchange_product_uom_qty(self):
             self.quantity = new_qty
 
+    @api.onchange('quantity')
+    def _onchange_quantity(self):
```
I suppose you can't fix this if the user set the qty through an rpc ?

lase-odoo replied:
I guess that we can try to modify the compute method instead !

adwid commented on addons/mrp/models/stock_move.py:
```diff
@@ -274,6 +274,12 @@ def _onchange_product_uom_qty(self):
+    @api.onchange('quantity')
+    def _onchange_quantity(self):
+        if self.raw_material_production_id and not self.manual_consumption and \
+           self.should_consume_qty != self.quantity:
```
"Always" use our float utils

lase-odoo replied:
Using the rounding precision of the `product_uom` of the stock move made at least 10 test fail because they use product without uom 🙃

I added a `precision_rounding = self.product_uom.rounding or 0.01`, If you prefer I can change all the tests..

adwid commented on addons/mrp/models/stock_move.py:
```diff
@@ -274,6 +274,12 @@ def _onchange_product_uom_qty(self):
+    def _onchange_quantity(self):
+        if self.raw_material_production_id and not self.manual_consumption and \
+           self.should_consume_qty != self.quantity:
+            self._origin.manual_consumption = True
```
`self._origin.manual_consumption`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;^^^^^^^
?

ticodoo commented:
Hello, I checked about this with the PO and this is the expected behavior now. The user is expected to either:

- set a non-zero `qty_producing` => manually change the component quantity => "Produce" (button)
- change the component to "Manual Consumption" in the BoM

So the user should be offered these options + this behavior should not be re-introduced (i.e. why the JS file was removed)

lase-odoo commented:
Hello @ticodoo, I am a bit puzzled. I discussed this use case with @mathieu-mgm last week and he told me that the "we need to make sure the demand is also 2 once the MO is done"
Did I miss Mathieu's point ?

mathieu-mgm commented:
Hello @lase-odoo I was talking about the 'to consume' (demand) quantity not the 'consumed' quantity. As mentioned if the issue is the consumed quantity than the issue must be solved with manual consumption on the bom. 
cc @ticodoo

adwid commented:
@mathieu-mgm 

Ola :smile: 

I don't really agree with closing the PR here, considering the written test/use case. We can be sure we will have some other tickets about this in the future, the use case is too common/simple. Whatever the configuration is, the user encodes a consumed quantity on his MO, _ticks the "Consumed" field_, validates the MO, and we just decide that he is wrong and force the consumed quantity ourselves...

Unfortunately, trying to set the `qty_producing` with a non-zero value is not that perfect either, because it depends on the order: if the user first encodes the consumed quantity, _ticks "Consumed"_, and then sets the producing qty, the onchange will just overwrite all encoded values :confused: 

So imho either it should work, (or we should prevent the user from changing the quantity)

mathieu-mgm commented:
hello @adwid I agree, the behaviour is not predictable enough. We could (as before) transform the move into a 'manual consumption' one once the checkbox on the MO move is ticked.

lase-odoo commented:
Hello @ticodoo, I have added a second test for the additional use case proposed by @adwid !
The same "fix" corrects the behavior appropriately in both situations. 
I also modified the onchange to only set manual consumption to the "picked" lines according to @mathieu-mgm 's suggestion :slightly_smiling_face:

mathieu-mgm commented:
@lase-odoo @ticodoo Just as a heads up, I would like to test this before we go ahead. Many thanks

mathieu-mgm commented:
@ticodoo looks good to me

ticodoo commented on addons/mrp/tests/test_manual_consumption.py:
```diff
@@ -230,3 +230,36 @@ def get_moves(mo):
         self.assertEqual(move_auto.manual_consumption, False)
         self.assertEqual(move_manual.manual_consumption, True)
+
+    def test_update_manual_consumption_00(self):
+        """
+        Check that the manual consumption is set to true when the quantity is manualy set.
+        """
+        mo_form = Form(self.env['mrp.production'])
+        mo_form.bom_id = self.bom_1
+        mo_form.product_qty = 1
+        mo = mo_form.save()
+        mo.action_confirm()
+        with Form(mo) as fmo:
+            with fmo.move_raw_ids.edit(0) as line_0:
+                line_0.quantity = 3.0
+                line_0.picked = True
+        mo.button_mark_done()
+        self.assertRecordValues(mo.move_raw_ids, [{'quantity': 3.0, 'picked': True}, {'quantity': 1.0, 'picked': True}])
+
+    def test_update_manual_consumption_01(self):
+        """
+        Check that the quantity of a raw line that is manually consumed is not updated
+        when the qty producing is changed and that others are.
+        """
+        mo_form = Form(self.env['mrp.production'])
+        mo_form.bom_id = self.bom_1
+        mo_form.product_qty = 1
+        mo = mo_form.save()
+        mo.action_confirm()
+        with Form(mo) as fmo:
+            with fmo.move_raw_ids.edit(0) as line_0:
+                line_0.quantity = 3.0
+                line_0.picked = True
+            fmo.qty_producing = 2.0
+        self.assertRecordValues(mo.move_raw_ids, [{'quantity': 3.0, 'picked': True}, {'quantity': 2.0, 'picked': True}])
```
I would make these tests a little more robust so that they're easier for people to understand in the future by doing the following:

- ensure the components are in stock (with `self.env['stock.quant']._update_available_quantity()`)
- add an assert that the BoM product quantities are reserved after the action_confirm (so others can see what the expected "To Consume" quantities are if it wasn't manually set)
-  add asserts that check that the `manual_consumption=True/False` for the moves are set as expected (in case someone changes the quantities for self.bom_1 which can make it so the test always passes in the end)

ticodoo commented on addons/mrp/models/stock_move.py:
```diff
@@ -274,6 +274,12 @@ def _onchange_product_uom_qty(self):
 
+    @api.onchange('quantity', 'product_uom', 'picked')
+    def _onchange_quantity(self):
+        if self.raw_material_production_id and not self.manual_consumption and self.picked and self.product_uom and \
```
`self.product_uom` is a required field :thinking: do you need to check that it's set?

lase-odoo replied:
Hello @ticodoo ! In a previous version of this fix, I did not include this check on the uom and then I realized that this breaks a lot of the mrp tests using forms:
https://runbot.odoo.com/runbot/build/62932160

The idea is that you can create "new" moves without products and hence with an empty "product_uom":
https://github.com/odoo/odoo/blob/663eb1fdb57f01f32ed791ce36478de2c65fb35a/addons/mrp/tests/test_bom.py#L1335-L1337

lase-odoo commented:
@ticodoo Thanks a lot for the review, the suggestions regarding the tests have been applied :slightly_smiling_face:
